#!/bin/bash
# IEKNGHER1ujm

# Load modules
module load bioinfo/bcftools

# Paths to main folders
DUMP="/home/dwragg/work/Analysis"
PIPE="/save/seqapipop/Scripts"
PERL="/save/dwragg/Scripts/Perl"
SEQA=/save/seqapipop/Data/Apis-mellifera
AMEL=/work/dwragg/Apis-mellifera

# Instead of using fixed value for upper DP limit, it can be replaced as follows:
# -- include 'DP>=9 & DP<(AVG(DP)*3)'

# SeqApiPop master sites [3,685,404] (just using JFM and OTH)
VCF_JFM=(${AMEL}/SeqApiPop/JFM/vcfs/*_clean.vcf.gz)
VCF_OTH=(${AMEL}/SeqApiPop/OTH/vcfs/*_clean.vcf.gz)
SAMPLES=(${VCF_JFM[@]} ${VCF_OTH[@]})
sites=${DUMP}/JFM-OTH-d9-1278.vcf.gz
bcftools merge ${SAMPLES[@]} \
  | bcftools view --types snps \
  --include 'DP>=9 & DP<(AVG(DP)*3)' \
  -G -M 2 -O z -o ${sites}
tabix -p vcf ${sites}
cp ${sites}* /save/seqapipop/Analysis/sites
rm ${sites}*

# Harpur master sites [10,444,626]
VCF_HARP=(${AMEL}/Harpur/vcfs/*_clean.vcf.gz)
sites=${DUMP}/HARP-d9-3113.vcf.gz
SAMPLES=(${VCF_HARP[@]})
bcftools merge ${SAMPLES[@]} \
  | bcftools view --types snps \
  --include 'DP>=9 & DP<(AVG(DP)*3)' \
  -G -M 2 -O z -o ${sites}
tabix -p vcf ${sites}
cp ${sites}* /save/seqapipop/Analysis/sites
rm ${sites}*

# Merged master sites
rm ${DUMP}/sites/*
ln -fs /save/seqapipop/Analysis/sites/* ${DUMP}/sites
sites=${DUMP}/sites/HARP-JFM-OTH.vcf.gz
bcftools merge -O z -o ${sites} \
  ${DUMP}/sites/HARP-d9-3113.vcf.gz \
  ${DUMP}/sites/JFM-OTH-d9-1278.vcf.gz
tabix -p vcf ${sites}
cp ${sites}* /save/seqapipop/Analysis/sites
rm ${sites}*
ln -fs /save/seqapipop/Analysis/sites/* ${DUMP}/sites


# ==============================================================================
# Sites
# ==============================================================================

# S1	Harpur-JFM-OTH.vcf.gz				[4,040,209]
# ------------------------------------------------------------------------------
# This list of sites has been generated using Harpur's Apis mellifera data with
# the JFM and OTH SeqApiPop data. SNPs are retained if they have a minimum
# allele frequency of 0.1, and a depth of coverage across all samples > 99 and
# < 3600 ( n * 2 * uDepth of Coverage).

# S2	HARP-JFM-OTH.vcf.gz				[11,242,975]
# ------------------------------------------------------------------------------
# This list of sites has been generated by merging the master sites identified
# in Harpur's Apis mellifera data and the JFM-OTH data, which have each
# independently be subjected to a depth of coverage filter (9 < DP < n*3*uDP)

# S3	Harpur-prune.vcf.gz				[38k]
# ------------------------------------------------------------------------------
# This is a pruned set of SNPs based on Harpur's AMCO reference samples, post-
# genotyping with the S2 (HARP-JFM-OTH) sites. Samples were subsequently
# merged (*_beagle.vcf.gz) for these sites, reduced to chromosomes 1-16 and
# filtered in Plink to retain those with weak LD (50, 10, 0.1), maf > 0.1, and
# SNP spacing > 5kb. Is used as a reference set for checking where sampels lie
# in relation to SNPs identified in the AMCO databset. The pipeline for this
# is located in "analysis-admixture-HarpurSNPs.sh"

# ==============================================================================
# Impute-Genotype
# ==============================================================================

# Prior to running this section, the sample names for each linaege need 
# assigning via the samples.sh script

# Generate file containing list of BAM files to operate on
# Sample-lineage assignments should be sources from samples.sh
printf "%s\n" "${VCF_A[@]}" > ${DUMP}/vcfs/lists/A_vcfs.list
printf "%s\n" "${BAM_A[@]}" > ${DUMP}/vcfs/lists/A_bams.list
printf "%s\n" "${VCF_M[@]}" > ${DUMP}/vcfs/lists/M_vcfs.list
printf "%s\n" "${BAM_M[@]}" > ${DUMP}/vcfs/lists/M_bams.list
printf "%s\n" "${VCF_C[@]}" > ${DUMP}/vcfs/lists/C_vcfs.list
printf "%s\n" "${BAM_C[@]}" > ${DUMP}/vcfs/lists/C_bams.list
printf "%s\n" "${VCF_O[@]}" > ${DUMP}/vcfs/lists/O_vcfs.list
printf "%s\n" "${BAM_O[@]}" > ${DUMP}/vcfs/lists/O_bams.list
printf "%s\n" "${VCF_JFM[@]}" > ${DUMP}/vcfs/lists/JFM_vcfs.list
printf "%s\n" "${BAM_JFM[@]}" > ${DUMP}/vcfs/lists/JFM_bams.list
printf "%s\n" "${VCF_OTH[@]}" > ${DUMP}/vcfs/lists/OTH_vcfs.list
printf "%s\n" "${BAM_OTH[@]}" > ${DUMP}/vcfs/lists/OTH_bams.list
printf "%s\n" "${VCF_BR[@]}" > ${DUMP}/vcfs/lists/BR_vcfs.list
printf "%s\n" "${BAM_BR[@]}" > ${DUMP}/vcfs/lists/BR_bams.list
printf "%s\n" "${VCF_COR[@]}" > ${DUMP}/vcfs/lists/COR_vcfs.list
printf "%s\n" "${BAM_COR[@]}" > ${DUMP}/vcfs/lists/COR_bams.list
printf "%s\n" "${VCF_OUE[@]}" > ${DUMP}/vcfs/lists/OUE_vcfs.list
printf "%s\n" "${BAM_OUE[@]}" > ${DUMP}/vcfs/lists/OUE_bams.list


# Output path and master site file
OUT=${DUMP}/vcfs/HARP-JFM-OTH-15012015
sites=${DUMP}/sites/HARP-JFM-OTH.vcf.gz

# Run each dataset (A, M, C, O, etc) on cluster (check ploidy -p)
qsub -q unlimitq -l mem=4G -l h_vmem=48G -pe parallel_smp 4 \
    -o ${DUMP}/logs \
    -e ${DUMP}/logs \
    ${PIPE}/imp-geno.sh -o ${OUT} -s ${sites} -p 1 \
      -v ${DUMP}/vcfs/lists/OTH_vcfs.list \
      -b ${DUMP}/vcfs/lists/OTH_bams.list \
      -n OTH

# Create diploids if necessary (for AMCO population analyses)
# <--------------------------- diploid-drones-notes.sh



# Diploid dataset for Admixture, etc.
bcftools merge -m snps -O z \
  ${OUT}/A_beagle.vcf.gz \
  ${OUT}/M_beagle.vcf.gz \
  ${OUT}/C_beagle.vcf.gz \
  ${OUT}/O_beagle.vcf.gz \
  ${OUT}/JFM_diploid.vcf.gz \
  ${OUT}/OTH_diploid.vcf.gz \
  -o ${OUT}/S2_AMCO-dipJFMOTH.vcf.gz
tabix -fp vcf ${OUT}/S2_AMCO-dipJFMOTH.vcf.gz

# JFMOTH_raw.vcf.gz == sites/JFM-OTH-d9-1278.vcf.gz
# This is the post-filtering set of variants for the data as described in the
# draft SeqApiPop manuscript

# Haploid dataset for Signatures of selection, etc.
bcftools merge -m snps -O z \
  ${OUT}/JFM_beagle.vcf.gz \
  ${OUT}/OTH_beagle.vcf.gz \
  -R ${OUT}/JFMOTH_raw.vcf.gz \
  -o ${OUT}/S2_AMCO-hapJFMOTH_beagle.vcf.gz
tabix -fp vcf ${OUT}/S2_AMCO-hapJFMOTH_beagle.vcf.gz

# Pre-imputation haploid dataset for Signatures of selection, etc.
bcftools merge -m snps -O z \
  ${OUT}/JFM_preImp.vcf.gz \
  ${OUT}/OTH_preImp.vcf.gz \
  -R ${OUT}/JFMOTH_raw.vcf.gz \
  -o ${OUT}/S2_hapJFMOTH_preImp.vcf.gz
tabix -fp vcf ${OUT}/S2_hapJFMOTH_preImp.vcf.gz


# ==============================================================================
# Miscellaneous notes
# ==============================================================================

# Reduce to just one dataset
#cd /save/seqapipop/Data/Apis-mellifera/Misc
#samples=(*)
# Retrieve sample IDs from samples.sh
#samples=(${A[@]} ${O[@]} ${M[@]} ${C[@]})
# Remove missing samples (domestic)
#samples=(${samples[@]/SRS549707/})
#samples=(${samples[@]/SRS549753/})
# Conert samples array to comma delimited string
#samples=$(printf ",%s" "${samples[@]}")
#samples=${samples:1}
#printf "%s\n" ${samples[@]} > ${DUMP}/samples/include.list

# Subset to new VCF
#bcftools view -m snps -O z \
#  -S ${DUMP}/samples/include.list \
#  -o ${DUMP}/vcfs/AMEL-4M/AMCO.vcf.gz \
#  ${DUMP}/vcfs/AMEL-4M/HARP_WALL.vcf.gz
#tabix -p vcf ${DUMP}/vcfs/AMEL-4M/AMCO.vcf.gz






